### ISSUE TITLE
Sistem za autonomno izbjegavanje sudara i prepoznavanje saobraćajnih znakova pri vožnji + Interaktivna simulacija za testiranje

### ISSUE THREAD

**Issue by SirBoi** [ftn-ai-lab/sc-2025#19]
## Tim
Stefan Lekić, SV 52/2022

## Asistent
Dragan Vidaković (@vdragan1993)

## Definicija problema
- Detekcija objekata u blizini automobila. Praćenje tih objekata kroz više frejmova i predikcija njihove pozicije u budućnosti. Računanje vjerovatnoće sudara sa objektom na osnovu njegove pozicije, smjera i brzine kretanja.
- Detekcija i identifikacija saobraćajnih znakova.
- Implementacija simuliranog okruženja koje dozvoljava korisniku da pomjera automobil kroz scenu sa pokretnim objektima.

## Motivacija problema
Ovakvi sistemi već nalaze primjenu u modernim automobilima opremljenim potrebnom tehnologijom (senzori, kontrola nad upravljanjem).
Razlog za implementaciju sistema u automobil je povećanje sigurnosti u saobraćaju, spasavanje ljudskih života i smanjenje materijalne štete.
Sistem je posebno koristan u slučajevima manjka pažnje vozača (izazvanog umorom, frustracijom, visokim temperaturama…) i neočekivanih medicinskih problema (onesvješćivanje, srčani/moždani udar…).

## Skup podataka

Potreban je skup podataka u svrhu treniranja modela za prepoznavanje i identifikaciju saobraćajnih znakova. Za ovo će se koristiti [German Traffic Sign Recognition Benchmark](https://benchmark.ini.rub.de/gtsrb_dataset.html#Pre-calculated_feature) dataset, koji raspoznaje 43 različita saobraćajna znaka i sastoji se od preko 50 000 slika.

## Metodologija

Sistem se sastoji od dva dijela:
- Simulacija (Unity, C#)
    > Predstavlja scenu sa automobilom i objektima (saobraćajnim znakovima, drugim automobilima, ljudima…).
    > Automobil je opremljen kamerom i LiDAR-om.
    > Kamera proizvodi RGB sliku.
    > LiDAR proizvodi set vrijednosti (a, d) (a = ugao posmatranja, d = udaljenost).
    > Automobil bilježi svoju trenutnu brzinu kretanja.

- Program za obradu (Python)
    > Pokreće server na koji se simulacija povezuje kao klijent.
    > Vrši obradu primljenih podataka.

Obrada podataka je zasnovana na sljedećim bibliotekama:
- Keras (TensorFlow) - Treniranje modela. Detekcija i identifikacija objekata.
- OpenCV - Obrada slike. Detekcija regiona. Vizuelizacija rezultata obrade.
- Pomoćne biblioteke - numpy, math…

Za obradu se koriste sljedeće metode:
- Transfer learning
- Sensor fusion
- Kalman filter

Ulazni podaci za obradu:
- Slika sa kamere
- LiDAR vrijednosti
- Brzina kretanja automobila

Izlazni podaci obrade:
- Nova ciljana brzina kretanja automobila

Ideja pipeline-a:
1. Očitavanje vrijednosti senzora
2. Prosljeđivanje podataka programu za obradu
3. Obrada podataka:
    3.1: Određivanje relevantnih dijelova slike pomoću LiDAR podataka
    3.2: Izdvajanje dijelova slike
    3.3: Priprema dijelova slike za dalju obradu
    3.4: Povezivanje izdvojenog objekta sa objektom iz prethodnog frejma
    3.5: Identifikacija saobraćajnog znaka/računanje smjera i brzine kretanja objekta
    3.6. Određivanje potrebe za usporavanjem automobila
4. Prosljeđivanje rezultata obrade simulaciji
5. Autonomno prilagođavanje brzine kretanja automobila

## Evaluacija

Model za prepoznavanje saobraćajnih znakova će biti evaluiran sljedećim metrikama:
- Accuracy
- Weighted precision
- Weighted recall
- Weighted F1 score
- Confusion matrix

Kompletan sistem će biti evaluiran uz pomoć predefinisanih test scenarija s automatizovanom vožnjom automobila. Svaki scenario će pokrivati više slučajeva u kojima sistem treba da se ponaša na određen način. Ispravnost sistema će biti evaluirana brojem ispravnih reakcija i _false positive_ rezultata za svaki scenario pojedinačno.

Osim toga, simulacija dozvoljava korisniku da ručno testira sistem u nekom od predefiniranih scenarija.

**Comment by vdragan1993** [comment]
Potrebno je proširiti Evaluaciju sa sledećim:
1. eksperimentalna postavka i metrike za evaluaciju modela za prepoznavanje saobraćajnih znakova.
2. evaluacija kompletnog pipeline-a tako da uključuje i metrike. U ovom slučaju je moguće samostalno definisati iste (npr. broj grešaka u određenom periodu i sl.).

**Comment by SirBoi** [comment]
Evaluacija je proširena traženim sadržajem.

**Comment by vdragan1993** [comment]
Tema odobrena.
Ostavite link ka GitHub repozitorijumu projekta.
Srećan rad.

**Comment by SirBoi** [comment]
https://github.com/SirBoi/autonomous-collision-avoidance-and-road-sign-detection-system
